language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -92,3 +95,155 @@ getBuiltinFix = do
         $ [ Apply () (unrollFunAB unroll2) $ Var () s
           , Var () x
           ]
+
+-- | A type that looks like a transformation.
+--
+-- > trans F G Q : F Q -> G Q
+trans :: Type TyName () -> Type TyName () -> Type TyName () -> Quote (Type TyName ())
+trans f g q = pure $ TyFun () (TyApp () f q) (TyApp () g q)
+
+-- | A type that looks like a natural transformation, sometimes written 'F ~> G'.
+--
+-- > natTrans F G : forall Q :: * . F Q -> G Q
+natTrans :: Type TyName () -> Type TyName () -> Quote (Type TyName ())
+natTrans f g = freshTyName () "Q" >>= \q -> TyForall () q (Type ()) <$> trans f g (TyVar () q)
+
+-- | A type that looks like a natural transformation to Id.
+--
+-- > natTransId F : forall Q :: * . F Q -> Q
+natTransId :: Type TyName () -> Quote (Type TyName ())
```

Comment:
I guess we need a helper module where we can put stuff like three functions from the above. We can do that later, though.